sql: possibly missing synchronization in DistSQLReceiver (or no longer valid assumption)

I ran into the following crash when stressing `TestDistributedQueryErrorIsRetriedLocally` with `vectorize=off` (this was on #98655, but I don't think that change is related)

```
* ERROR: a SQL panic has occurred while executing the following statement:
* SELECT k FROM test.foo
*
*
* ERROR: a panic has occurred!
* can't call AddRow after having set error: connection refused: n2
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:952
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Wait
*   | 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:614
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).beforeAdd
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:223
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).AddRow
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:241
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).Push
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1479
*   | github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*copyingRowReceiver).Push
*   | 	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:486
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:197
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:726
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:901
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1924
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1639
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1642
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2291
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1839
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1070
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:144
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3237
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:143
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2223
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2228
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2145
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:954
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:247
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:999
* Wraps: (2) while executing: SELECT _ FROM _
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:952
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Wait
*   | 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:614
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | [...repeated from below...]
* Wraps: (4) assertion failure
* Wraps: (5) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).beforeAdd
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:223
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).AddRow
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:241
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).Push
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1479
*   | github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*copyingRowReceiver).Push
*   | 	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:486
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:197
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:726
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:901
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1924
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1639
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1642
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2291
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1839
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1070
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:144
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3237
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:143
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2223
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2228
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2145
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:954
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommands
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:247
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveImpl.func3
*   | 	github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:999
*   | runtime.goexit
*   | 	src/runtime/asm_arm64.s:1172
* Wraps: (6) can't call AddRow after having set error
* Wraps: (7) connection refused: n2
*   | -- cause hidden behind barrier
*   | connection refused: n2
*   | (1) tags: [n1,client=127.0.0.1:50834,hostssl,user=root,received-error]
*   | Wraps: (2) tags: [n2,f845ed2a2,distsql.stmt=SELECT k FROM test.foo,distsql.gateway=1,distsql.txn=c9e50183-3b5f-4ef9-9e46-3a07c59a0caf,sent-error=]
*   | Wraps: (3)
*   |   | (opaque error wrapper)
*   |   | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
*   |   | reportable 0:
*   |   |
*   |   | github.com/cockroachdb/cockroach/pkg/sql.TestDistributedQueryErrorIsRetriedLocally.func1
*   |   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running_test.go:989
*   |   | github.com/cockroachdb/cockroach/pkg/sql.TestDistributedQueryErrorIsRetriedLocally.func2
*   |   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running_test.go:1022
*   |   | github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow.func1
*   |   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:683
*   |   | github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow
*   |   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:689
*   |   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1
*   |   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/darwin_arm64-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:602
*   |   | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1
*   |   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97
*   |   | google.golang.org/grpc.getChainUnaryHandler.func1
*   |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163
*   |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
*   |   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169
*   |   | google.golang.org/grpc.getChainUnaryHandler.func1
*   |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163
*   |   | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
*   |   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:105
*   |   | google.golang.org/grpc.getChainUnaryHandler.func1
*   |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163
*   |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
*   |   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:136
*   |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
*   |   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:336
*   |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
*   |   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:134
*   |   | google.golang.org/grpc.chainUnaryInterceptors.func1
*   |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154
*   |   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler
*   |   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/darwin_arm64-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:604
*   |   | google.golang.org/grpc.(*Server).processUnaryRPC
*   |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1336
*   |   | google.golang.org/grpc.(*Server).handleStream
*   |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1704
*   |   | google.golang.org/grpc.(*Server).serveStreams.func1.2
*   |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:965
*   |   | runtime.goexit
*   |   | 	src/runtime/asm_arm64.s:1172
*   | Wraps: (4) connection refused: n2
*   | Error types: (1) *contexttags.withContext (2) *contexttags.withContext (3) *errbase.opaqueWrapper (4) *errutil.leafError
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *assert.withAssertionFailure (5) *withstack.withStack (6) *errutil.withPrefix (7) *barriers.barrierErr
*
panic: can't call AddRow after having set error: connection refused: n2 [recovered]
	panic: can't call AddRow after having set error: connection refused: n2 [recovered]
	panic: can't call AddRow after having set error: connection refused: n2
```
Currently `DistSQLReceiver.resultWriterMu` only protects `SetError` and `Err` calls, but it seems possible for `AddRow` to be called by the main goroutine concurrently with distsql runner goroutine setting the error when `SetupFlow` RPC fails. Perhaps we simply need to remove the assertion since it no longer holds true.

Jira issue: CRDB-32175

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -460,7 +460,7 @@ func (dsp *DistSQLPlanner) setupFlows(
 	// First, set up the flow on this node.
 	setupReq.Flow = *flows[thisNodeID]
 	var batchReceiver execinfra.BatchReceiver
-	if recv.resultWriterMu.batch != nil {
+	if recv.batchWriter != nil {
 		// Use the DistSQLReceiver as an execinfra.BatchReceiver only if the
 		// former has the corresponding writer set.
 		batchReceiver = recv
@@ -590,6 +590,8 @@ func (dsp *DistSQLPlanner) setupFlows(
 	// since flow.Cleanup() is the last thing before DistSQLPlanner.Run returns
 	// at which point the rowResultWriter is no longer protected by the mutex of
 	// the DistSQLReceiver.
+	// TODO(yuzefovich): think through this comment - we no longer have mutex
+	// protection in place, so perhaps this can be simplified.
 	cleanupCalledMu := struct {
 		syncutil.Mutex
 		called bool
@@ -623,14 +625,10 @@ func (dsp *DistSQLPlanner) setupFlows(
 						// so there is nothing to do.
 						return
 					}
-					// First, we update the DistSQL receiver with the error to
-					// be returned to the client eventually.
-					//
-					// In order to not protect DistSQLReceiver.status with a
-					// mutex, we do not update the status here and, instead,
-					// rely on the DistSQLReceiver detecting the error the next
-					// time an object is pushed into it.
-					recv.setErrorWithoutStatusUpdate(res.err, true /* willDeferStatusUpdate */)
+					// First, we send the error to the DistSQL receiver to be
+					// returned to the client eventually (which will happen on
+					// the next Push or PushBatch call).
+					recv.concurrentErrorCh <- res.err
 					// Now explicitly cancel the local flow.
 					flow.Cancel()
 				}()
@@ -835,9 +833,11 @@ func (dsp *DistSQLPlanner) Run(
 		// We ended up planning everything locally, regardless of whether we
 		// intended to distribute or not.
 		localState.IsLocal = true
+		// Concurrent errors are only possible during the distributed execution.
+		recv.skipConcurrentErrorCheck = true
 	} else {
 		defer func() {
-			if recv.getError() != nil {
+			if recv.resultWriter.Err() != nil {
 				// The execution of this query encountered some error, so we
 				// will eagerly cancel all flows running on the remote nodes
 				// because they are now dead.
@@ -916,21 +916,25 @@ func (dsp *DistSQLPlanner) Run(
 type DistSQLReceiver struct {
 	ctx context.Context
 
-	resultWriterMu struct {
-		// Mutex only protects SetError() and Err() methods of the
-		// rowResultWriter.
-		syncutil.Mutex
-		// These two interfaces refer to the same object, but batch might be
-		// unset (row is always set). These are used to send the results to.
-		row   rowResultWriter
-		batch batchResultWriter
-	}
-	// updateStatus, if true, indicates that a concurrent goroutine has set an
-	// error on the rowResultWriter without updating status, so the main
-	// goroutine needs to update the status.
-	updateStatus atomic.Bool
-	status       execinfra.ConsumerStatus
+	// These two interfaces refer to the same object, but batchWriter might be
+	// unset (resultWriter is always set). These are used to send the results
+	// to.
+	resultWriter rowResultWriter
+	batchWriter  batchResultWriter
 
+	// concurrentErrorCh is a buffered channel that allows for concurrent
+	// goroutines to tell the main execution goroutine that there is an error.
+	// The main goroutine will read from this channel on the next call to Push
+	// or PushBatch.
+	//
+	// This channel is needed since rowResultWriter is not thread-safe and will
+	// only be sent on during distributed plan execution.
+	concurrentErrorCh chan error
+	// skipConcurrentErrorCheck is set when concurrent errors aren't possible,
+	// so there is no need to poll concurrentErrorCh.
+	skipConcurrentErrorCheck bool
+
+	status   execinfra.ConsumerStatus
 	stmtType tree.StatementReturnType
 
 	// outputTypes are the types of the result columns produced by the plan.
@@ -1220,15 +1224,18 @@ func MakeDistSQLReceiver(
 	}
 	*r = DistSQLReceiver{
 		ctx:          consumeCtx,
-		cleanup:      cleanup,
-		rangeCache:   rangeCache,
-		txn:          txn,
-		clockUpdater: clockUpdater,
-		stmtType:     stmtType,
-		tracing:      tracing,
+		resultWriter: resultWriter,
+		batchWriter:  batchWriter,
+		// At the time of writing, there is only one concurrent goroutine that
+		// might send at most one error.
+		concurrentErrorCh: make(chan error, 1),
+		cleanup:           cleanup,
+		rangeCache:        rangeCache,
+		txn:               txn,
+		clockUpdater:      clockUpdater,
+		stmtType:          stmtType,
+		tracing:           tracing,
 	}
-	r.resultWriterMu.row = resultWriter
-	r.resultWriterMu.batch = batchWriter
 	return r
 }
 
@@ -1236,8 +1243,11 @@ func MakeDistSQLReceiver(
 // executing the plan - that encountered an error when run in the distributed
 // fashion - locally.
 func (r *DistSQLReceiver) resetForLocalRerun(stats topLevelQueryStats) {
-	r.resultWriterMu.row.SetError(nil)
-	r.updateStatus.Store(false)
+	r.resultWriter.SetError(nil)
+	// Note that DistSQLPlanner.Run will also set r.skipConcurrentErrorCheck to
+	// true, so r.concurrentErrorCh won't be pulled during the local rerun. Here
+	// we do so for clarity.
+	r.skipConcurrentErrorCheck = true
 	r.status = execinfra.NeedMoreRows
 	r.dataPushed = false
 	r.closed = false
@@ -1258,36 +1268,27 @@ func (r *DistSQLReceiver) Release() {
 func (r *DistSQLReceiver) clone() *DistSQLReceiver {
 	ret := receiverSyncPool.Get().(*DistSQLReceiver)
 	*ret = DistSQLReceiver{
-		ctx:          r.ctx,
-		cleanup:      func() {},
-		rangeCache:   r.rangeCache,
-		txn:          r.txn,
-		clockUpdater: r.clockUpdater,
-		stmtType:     tree.Rows,
-		tracing:      r.tracing,
+		ctx:               r.ctx,
+		concurrentErrorCh: make(chan error, 1),
+		cleanup:           func() {},
+		rangeCache:        r.rangeCache,
+		txn:               r.txn,
+		clockUpdater:      r.clockUpdater,
+		stmtType:          tree.Rows,
+		tracing:           r.tracing,
 	}
 	return ret
 }
 
-// getError returns the error stored in the rowResultWriter (if any).
-func (r *DistSQLReceiver) getError() error {
-	r.resultWriterMu.Lock()
-	defer r.resultWriterMu.Unlock()
-	return r.resultWriterMu.row.Err()
-}
-
-// setErrorWithoutStatusUpdate sets the error in the rowResultWriter but does
-// **not** update the status of the DistSQLReceiver. willDeferStatusUpdate
-// indicates whether the main goroutine should update the status the next time
-// it pushes something into the DistSQLReceiver.
+// SetError passes the given error to the resultWriter and updates the status of
+// the DistSQLReceiver accordingly.
 //
-// NOTE: consider using SetError() instead.
-func (r *DistSQLReceiver) setErrorWithoutStatusUpdate(err error, willDeferStatusUpdate bool) {
-	r.resultWriterMu.Lock()
-	defer r.resultWriterMu.Unlock()
+// Note that this method is **not** concurrency safe and needs to be called only
+// from the main execution goroutine.
+func (r *DistSQLReceiver) SetError(err error) {
 	// Check if the error we just received should take precedence over a
 	// previous error (if any).
-	if kvpb.ErrPriority(err) > kvpb.ErrPriority(r.resultWriterMu.row.Err()) {
+	if kvpb.ErrPriority(err) > kvpb.ErrPriority(r.resultWriter.Err()) {
 		if r.txn != nil {
 			if retryErr := (*kvpb.UnhandledRetryableError)(nil); errors.As(err, &retryErr) {
 				// Update the txn in response to remote errors. In the
@@ -1306,54 +1307,39 @@ func (r *DistSQLReceiver) setErrorWithoutStatusUpdate(err error, willDeferStatus
 				}
 			}
 		}
-		r.resultWriterMu.row.SetError(err)
-		r.updateStatus.Store(willDeferStatusUpdate)
+		r.resultWriter.SetError(err)
+		// If we encountered an error, we will transition to draining unless we
+		// were canceled.
+		if r.ctx.Err() != nil {
+			log.VEventf(r.ctx, 1, "encountered error (transitioning to shutting down): %v", r.ctx.Err())
+			r.status = execinfra.ConsumerClosed
+		} else {
+			log.VEventf(r.ctx, 1, "encountered error (transitioning to draining): %v", err)
+			r.status = execinfra.DrainRequested
+		}
 	}
 }
 
-// updateStatusAfterError updates the status of the DistSQLReceiver after it
-// has received an error.
-func (r *DistSQLReceiver) updateStatusAfterError(err error) {
-	// If we encountered an error, we will transition to draining unless we were
-	// canceled.
-	if r.ctx.Err() != nil {
-		log.VEventf(r.ctx, 1, "encountered error (transitioning to shutting down): %v", r.ctx.Err())
-		r.status = execinfra.ConsumerClosed
-	} else {
-		log.VEventf(r.ctx, 1, "encountered error (transitioning to draining): %v", err)
-		r.status = execinfra.DrainRequested
-	}
-}
-
-// SetError provides a convenient way for a client to pass in an error, thus
-// pretending that a query execution error happened. The error is passed along
-// to the resultWriter.
-//
-// The status of DistSQLReceiver is updated accordingly.
-func (r *DistSQLReceiver) SetError(err error) {
-	r.setErrorWithoutStatusUpdate(err, false /* willDeferStatusUpdate */)
-	r.updateStatusAfterError(err)
-}
-
-// checkConcurrentError sets the status if an error has been set by another
-// goroutine that did not also update the status.
+// checkConcurrentError checks whether a concurrent goroutine encountered an
+// error that should trigger the shutdown of the execution pipeline. The status
+// of the DistSQLReceiver is updated accordingly.
 func (r *DistSQLReceiver) checkConcurrentError() {
-	if r.status != execinfra.NeedMoreRows || !r.updateStatus.Load() {
+	if r.skipConcurrentErrorCheck || r.status != execinfra.NeedMoreRows {
 		// If the status already is not NeedMoreRows, then it doesn't matter if
 		// there was a concurrent error set.
 		return
 	}
-	previousErr := r.getError()
-	if previousErr == nil {
-		previousErr = errors.AssertionFailedf("unexpectedly updateStatus is set but there is no error")
+	select {
+	case err := <-r.concurrentErrorCh:
+		r.SetError(err)
+	default:
 	}
-	r.updateStatusAfterError(previousErr)
 }
 
 // pushMeta takes in non-empty metadata object and pushes it to the result
 // writer. Possibly updated status is returned.
 func (r *DistSQLReceiver) pushMeta(meta *execinfrapb.ProducerMetadata) execinfra.ConsumerStatus {
-	if metaWriter, ok := r.resultWriterMu.row.(MetadataResultWriter); ok {
+	if metaWriter, ok := r.resultWriter.(MetadataResultWriter); ok {
 		metaWriter.AddMeta(r.ctx, meta)
 		if meta.Err == nil {
 			r.dataPushed = true
@@ -1458,7 +1444,7 @@ func (r *DistSQLReceiver) Push(
 		// We only need the row count. planNodeToRowSource is set up to handle
 		// ensuring that the last stage in the pipeline will return a single-column
 		// row with the row count in it, so just grab that and exit.
-		r.resultWriterMu.row.SetRowsAffected(r.ctx, n)
+		r.resultWriter.SetRowsAffected(r.ctx, n)
 		return r.status
 	}
 
@@ -1509,7 +1495,7 @@ func (r *DistSQLReceiver) Push(
 		}
 	}
 	r.tracing.TraceExecRowsResult(r.ctx, r.row)
-	if commErr := r.resultWriterMu.row.AddRow(r.ctx, r.row); commErr != nil {
+	if commErr := r.resultWriter.AddRow(r.ctx, r.row); commErr != nil {
 		r.handleCommErr(commErr)
 	}
 	r.dataPushed = true
@@ -1543,7 +1529,7 @@ func (r *DistSQLReceiver) PushBatch(
 		// We only need the row count. planNodeToRowSource is set up to handle
 		// ensuring that the last stage in the pipeline will return a single-column
 		// row with the row count in it, so just grab that and exit.
-		r.resultWriterMu.row.SetRowsAffected(r.ctx, int(batch.ColVec(0).Int64()[0]))
+		r.resultWriter.SetRowsAffected(r.ctx, int(batch.ColVec(0).Int64()[0]))
 		return r.status
 	}
 
@@ -1565,7 +1551,7 @@ func (r *DistSQLReceiver) PushBatch(
 		panic("unsupported exists mode for PushBatch")
 	}
 	r.tracing.TraceExecBatchResult(r.ctx, batch)
-	if commErr := r.resultWriterMu.batch.AddBatch(r.ctx, batch); commErr != nil {
+	if commErr := r.batchWriter.AddBatch(r.ctx, batch); commErr != nil {
 		r.handleCommErr(commErr)
 	}
 	r.dataPushed = true
@@ -1693,7 +1679,7 @@ func (dsp *DistSQLPlanner) PlanAndRunAll(
 		}
 	}
 
-	if recv.commErr != nil || recv.getError() != nil {
+	if recv.commErr != nil || recv.resultWriter.Err() != nil {
 		return recv.commErr
 	}
 
@@ -1810,7 +1796,7 @@ func (dsp *DistSQLPlanner) planAndRunSubquery(
 
 	// TODO(yuzefovich): consider implementing batch receiving result writer.
 	subqueryRowReceiver := NewRowResultWriter(&rows)
-	subqueryRecv.resultWriterMu.row = subqueryRowReceiver
+	subqueryRecv.resultWriter = subqueryRowReceiver
 	subqueryPlans[planIdx].started = true
 	finishedSetupFn, cleanup := getFinishedSetupFn(planner)
 	defer cleanup()
@@ -1962,7 +1948,7 @@ func (dsp *DistSQLPlanner) PlanAndRun(
 		// was encountered or not.
 		return
 	}
-	if distributedErr := recv.getError(); distributedErr != nil &&
+	if distributedErr := recv.resultWriter.Err(); distributedErr != nil &&
 		distributedQueryRerunAsLocalEnabled.Get(&dsp.st.SV) {
 		// If we had a distributed plan which resulted in an error, we want to
 		// retry this query as local in some cases. In particular, this retry
@@ -2007,7 +1993,7 @@ func (dsp *DistSQLPlanner) PlanAndRun(
 		telemetry.Inc(sqltelemetry.DistributedErrorLocalRetryAttempt)
 		dsp.distSQLSrv.ServerConfig.Metrics.DistErrorLocalRetryAttempts.Inc(1)
 		defer func() {
-			if recv.getError() != nil {
+			if recv.resultWriter.Err() != nil {
 				telemetry.Inc(sqltelemetry.DistributedErrorLocalRetryFailure)
 				dsp.distSQLSrv.ServerConfig.Metrics.DistErrorLocalRetryFailures.Inc(1)
 			}
@@ -2277,12 +2263,12 @@ func (dsp *DistSQLPlanner) planAndRunPostquery(
 	defer postqueryRecv.Release()
 	defer addTopLevelQueryStats(&postqueryRecv.stats)
 	postqueryResultWriter := &errOnlyResultWriter{}
-	postqueryRecv.resultWriterMu.row = postqueryResultWriter
-	postqueryRecv.resultWriterMu.batch = postqueryResultWriter
+	postqueryRecv.resultWriter = postqueryResultWriter
+	postqueryRecv.batchWriter = postqueryResultWriter
 	finishedSetupFn, cleanup := getFinishedSetupFn(planner)
 	defer cleanup()
 	dsp.Run(ctx, postqueryPlanCtx, planner.txn, postqueryPhysPlan, postqueryRecv, evalCtx, finishedSetupFn)
-	return postqueryRecv.getError()
+	return postqueryRecv.resultWriter.Err()
 }
 
 // planAndRunChecksInParallel executes all checkPlans in parallel. The function

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -625,7 +625,12 @@ func (dsp *DistSQLPlanner) setupFlows(
 					}
 					// First, we update the DistSQL receiver with the error to
 					// be returned to the client eventually.
-					recv.SetError(res.err)
+					//
+					// In order to not protect DistSQLReceiver.status with a
+					// mutex, we do not update the status here and, instead,
+					// rely on the DistSQLReceiver detecting the error the next
+					// time an object is pushed into it.
+					recv.setErrorWithoutStatusUpdate(res.err, true /* willDeferStatusUpdate */)
 					// Now explicitly cancel the local flow.
 					flow.Cancel()
 				}()
@@ -920,9 +925,11 @@ type DistSQLReceiver struct {
 		row   rowResultWriter
 		batch batchResultWriter
 	}
-	// status is execinfra.ConsumerStatus of the receiver. It should not be
-	// accessed directly - use getStatus() and setStatus() helpers instead.
-	status atomic.Uint32
+	// updateStatus, if true, indicates that a concurrent goroutine has set an
+	// error on the rowResultWriter without updating status, so the main
+	// goroutine needs to update the status.
+	updateStatus atomic.Bool
+	status       execinfra.ConsumerStatus
 
 	stmtType tree.StatementReturnType
 
@@ -991,14 +998,6 @@ type DistSQLReceiver struct {
 		// Possibly updated arguments are returned.
 		pushCallback func(rowenc.EncDatumRow, coldata.Batch, *execinfrapb.ProducerMetadata) (rowenc.EncDatumRow, coldata.Batch, *execinfrapb.ProducerMetadata)
 	}
-}
-
-func (r *DistSQLReceiver) getStatus() execinfra.ConsumerStatus {
-	return execinfra.ConsumerStatus(r.status.Load())
-}
-
-func (r *DistSQLReceiver) setStatus(s execinfra.ConsumerStatus) {
-	r.status.Store(uint32(s))
 }
 
 // rowResultWriter is a subset of CommandResult to be used with the
@@ -1238,7 +1237,8 @@ func MakeDistSQLReceiver(
 // fashion - locally.
 func (r *DistSQLReceiver) resetForLocalRerun(stats topLevelQueryStats) {
 	r.resultWriterMu.row.SetError(nil)
-	r.setStatus(execinfra.NeedMoreRows)
+	r.updateStatus.Store(false)
+	r.status = execinfra.NeedMoreRows
 	r.dataPushed = false
 	r.closed = false
 	r.stats = stats
@@ -1276,12 +1276,13 @@ func (r *DistSQLReceiver) getError() error {
 	return r.resultWriterMu.row.Err()
 }
 
-// SetError provides a convenient way for a client to pass in an error, thus
-// pretending that a query execution error happened. The error is passed along
-// to the resultWriter.
+// setErrorWithoutStatusUpdate sets the error in the rowResultWriter but does
+// **not** update the status of the DistSQLReceiver. willDeferStatusUpdate
+// indicates whether the main goroutine should update the status the next time
+// it pushes something into the DistSQLReceiver.
 //
-// The status of DistSQLReceiver is updated accordingly.
-func (r *DistSQLReceiver) SetError(err error) {
+// NOTE: consider using SetError() instead.
+func (r *DistSQLReceiver) setErrorWithoutStatusUpdate(err error, willDeferStatusUpdate bool) {
 	r.resultWriterMu.Lock()
 	defer r.resultWriterMu.Unlock()
 	// Check if the error we just received should take precedence over a
@@ -1306,16 +1307,47 @@ func (r *DistSQLReceiver) SetError(err error) {
 			}
 		}
 		r.resultWriterMu.row.SetError(err)
+		r.updateStatus.Store(willDeferStatusUpdate)
 	}
+}
+
+// updateStatusAfterError updates the status of the DistSQLReceiver after it
+// has received an error.
+func (r *DistSQLReceiver) updateStatusAfterError(err error) {
 	// If we encountered an error, we will transition to draining unless we were
 	// canceled.
 	if r.ctx.Err() != nil {
 		log.VEventf(r.ctx, 1, "encountered error (transitioning to shutting down): %v", r.ctx.Err())
-		r.setStatus(execinfra.ConsumerClosed)
+		r.status = execinfra.ConsumerClosed
 	} else {
 		log.VEventf(r.ctx, 1, "encountered error (transitioning to draining): %v", err)
-		r.setStatus(execinfra.DrainRequested)
+		r.status = execinfra.DrainRequested
 	}
+}
+
+// SetError provides a convenient way for a client to pass in an error, thus
+// pretending that a query execution error happened. The error is passed along
+// to the resultWriter.
+//
+// The status of DistSQLReceiver is updated accordingly.
+func (r *DistSQLReceiver) SetError(err error) {
+	r.setErrorWithoutStatusUpdate(err, false /* willDeferStatusUpdate */)
+	r.updateStatusAfterError(err)
+}
+
+// checkConcurrentError sets the status if an error has been set by another
+// goroutine that did not also update the status.
+func (r *DistSQLReceiver) checkConcurrentError() {
+	if r.status != execinfra.NeedMoreRows || !r.updateStatus.Load() {
+		// If the status already is not NeedMoreRows, then it doesn't matter if
+		// there was a concurrent error set.
+		return
+	}
+	previousErr := r.getError()
+	if previousErr == nil {
+		previousErr = errors.AssertionFailedf("unexpectedly updateStatus is set but there is no error")
+	}
+	r.updateStatusAfterError(previousErr)
 }
 
 // pushMeta takes in non-empty metadata object and pushes it to the result
@@ -1362,7 +1394,7 @@ func (r *DistSQLReceiver) pushMeta(meta *execinfrapb.ProducerMetadata) execinfra
 	}
 	// Release the meta object. It is unsafe for use after this call.
 	meta.Release()
-	return r.getStatus()
+	return r.status
 }
 
 // handleCommErr handles the communication error (the one returned when
@@ -1373,12 +1405,12 @@ func (r *DistSQLReceiver) handleCommErr(commErr error) {
 	// client (that's why we don't set the error on the resultWriter).
 	if errors.Is(commErr, ErrLimitedResultClosed) {
 		log.VEvent(r.ctx, 1, "encountered ErrLimitedResultClosed (transitioning to draining)")
-		r.setStatus(execinfra.DrainRequested)
+		r.status = execinfra.DrainRequested
 	} else if errors.Is(commErr, errIEResultChannelClosed) {
 		log.VEvent(r.ctx, 1, "encountered errIEResultChannelClosed (transitioning to draining)")
-		r.setStatus(execinfra.DrainRequested)
+		r.status = execinfra.DrainRequested
 	} else if errors.Is(commErr, ErrPortalLimitHasBeenReached) {
-		r.setStatus(execinfra.SwitchToAnotherPortal)
+		r.status = execinfra.SwitchToAnotherPortal
 	} else {
 		// Set the error on the resultWriter to notify the consumer about
 		// it. Most clients don't care to differentiate between
@@ -1407,17 +1439,18 @@ func (r *DistSQLReceiver) handleCommErr(commErr error) {
 func (r *DistSQLReceiver) Push(
 	row rowenc.EncDatumRow, meta *execinfrapb.ProducerMetadata,
 ) execinfra.ConsumerStatus {
+	r.checkConcurrentError()
 	if r.testingKnobs.pushCallback != nil {
 		row, _, meta = r.testingKnobs.pushCallback(row, nil /* batch */, meta)
 	}
 	if meta != nil {
 		return r.pushMeta(meta)
 	}
-	if r.ctx.Err() != nil && r.getStatus() != execinfra.ConsumerClosed {
+	if r.ctx.Err() != nil && r.status != execinfra.ConsumerClosed {
 		r.SetError(r.ctx.Err())
 	}
-	if status := r.getStatus(); status != execinfra.NeedMoreRows {
-		return status
+	if r.status != execinfra.NeedMoreRows {
+		return r.status
 	}
 
 	if r.stmtType != tree.Rows {
@@ -1426,7 +1459,7 @@ func (r *DistSQLReceiver) Push(
 		// ensuring that the last stage in the pipeline will return a single-column
 		// row with the row count in it, so just grab that and exit.
 		r.resultWriterMu.row.SetRowsAffected(r.ctx, n)
-		return r.getStatus()
+		return r.status
 	}
 
 	ensureDecodedRow := func() error {
@@ -1446,11 +1479,11 @@ func (r *DistSQLReceiver) Push(
 	if r.isTenantExplainAnalyze {
 		if err := ensureDecodedRow(); err != nil {
 			r.SetError(err)
-			return r.getStatus()
+			return r.status
 		}
 		if len(r.row) != len(r.outputTypes) {
 			r.SetError(errors.Errorf("expected number of columns and output types to be the same"))
-			return r.getStatus()
+			return r.status
 		}
 		if r.egressCounter == nil {
 			r.egressCounter = NewTenantNetworkEgressCounter()
@@ -1460,7 +1493,7 @@ func (r *DistSQLReceiver) Push(
 
 	if r.discardRows {
 		// Discard rows.
-		return r.getStatus()
+		return r.status
 	}
 
 	if r.existsMode {
@@ -1468,11 +1501,11 @@ func (r *DistSQLReceiver) Push(
 		// row is pushed or not, so the contents do not matter.
 		r.row = []tree.Datum{}
 		log.VEvent(r.ctx, 2, `a row is pushed in "exists" mode, so transition to draining`)
-		r.setStatus(execinfra.DrainRequested)
+		r.status = execinfra.DrainRequested
 	} else {
 		if err := ensureDecodedRow(); err != nil {
 			r.SetError(err)
-			return r.getStatus()
+			return r.status
 		}
 	}
 	r.tracing.TraceExecRowsResult(r.ctx, r.row)
@@ -1480,29 +1513,30 @@ func (r *DistSQLReceiver) Push(
 		r.handleCommErr(commErr)
 	}
 	r.dataPushed = true
-	return r.getStatus()
+	return r.status
 }
 
 // PushBatch is part of the execinfra.BatchReceiver interface.
 func (r *DistSQLReceiver) PushBatch(
 	batch coldata.Batch, meta *execinfrapb.ProducerMetadata,
 ) execinfra.ConsumerStatus {
+	r.checkConcurrentError()
 	if r.testingKnobs.pushCallback != nil {
 		_, batch, meta = r.testingKnobs.pushCallback(nil /* row */, batch, meta)
 	}
 	if meta != nil {
 		return r.pushMeta(meta)
 	}
-	if r.ctx.Err() != nil && r.getStatus() != execinfra.ConsumerClosed {
+	if r.ctx.Err() != nil && r.status != execinfra.ConsumerClosed {
 		r.SetError(r.ctx.Err())
 	}
-	if status := r.getStatus(); status != execinfra.NeedMoreRows {
-		return status
+	if r.status != execinfra.NeedMoreRows {
+		return r.status
 	}
 
 	if batch.Length() == 0 {
 		// Nothing to do on the zero-length batch.
-		return r.getStatus()
+		return r.status
 	}
 
 	if r.stmtType != tree.Rows {
@@ -1510,7 +1544,7 @@ func (r *DistSQLReceiver) PushBatch(
 		// ensuring that the last stage in the pipeline will return a single-column
 		// row with the row count in it, so just grab that and exit.
 		r.resultWriterMu.row.SetRowsAffected(r.ctx, int(batch.ColVec(0).Int64()[0]))
-		return r.getStatus()
+		return r.status
 	}
 
 	if r.isTenantExplainAnalyze {
@@ -1522,7 +1556,7 @@ func (r *DistSQLReceiver) PushBatch(
 
 	if r.discardRows {
 		// Discard rows.
-		return r.getStatus()
+		return r.status
 	}
 
 	if r.existsMode {
@@ -1535,7 +1569,7 @@ func (r *DistSQLReceiver) PushBatch(
 		r.handleCommErr(commErr)
 	}
 	r.dataPushed = true
-	return r.getStatus()
+	return r.status
 }
 
 var (

